import edu.ucr.cs.riple.taint.ucrtainting.qual.*;

class Foo {

  String field = "";
  Bar bar = new Bar();

  void bar(String x, @RUntainted String y, boolean b) {
    String localVar = x;
    final String finalLocalVar = x;
    // :: error: assignment
    @RUntainted String c = field;
    // :: error: assignment
    c = x;
    // :: error: assignment
    c = b ? x : y;
    // :: error: assignment
    c = x + y;
    // :: error: assignment
    c = b ? x + y : localVar;
    if (b) {
      // :: error: assignment
      c = bar.getField();
    } else {
      // :: error: assignment
      c = bar.field;
      // :: error: assignment
      c = x;
      // :: error: assignment
      c = localVar;
      // :: error: assignment
      c = bar.baz.field;
    }
    class LocalClass {
      @RUntainted String field;
      String f2;

      void foo() {
        // :: error: assignment
        field = finalLocalVar;
        // :: error: assignment
        field = f2;
        String localVar = f2;
        class LocalInnerClass {
          // :: error: assignment
          @RUntainted String baz = localVar;
        }
      }
    }
    // :: error: assignment
    c = bar.staticF;
  }

  public Object inherit(Object param) {
    return param;
  }
}
