package foo.bar;

import edu.ucr.cs.riple.taint.ucrtainting.qual.*;

public class Test {
  //
  //    public List<String> getURIList(final String key, final List<String> defaultValue) {
  //        // :: error: return
  //        // :: error: argument
  //        return getList(String.class, key, defaultValue);
  //    }
  //    public <T> List<T> getList(final Class<T> cls, final String key, final List<T> defaultValue)
  // {
  //        return null;
  //    }
}
